packages/no-trapping-shim/src/no-trapping-pony.js

```diff
@@ -0,0 +1,278 @@
+     * @param {any[]} rest
+     */
+    const trapPlus = freeze((target, ...rest) => {
+      if (isNoTrappingInternal(target, true)) {
+        defineProperty(handlerPlus, trapName, {
+          value: undefined,
+          writable: false,
+          enumerable: true,
+          configurable: false,
+        });
+      } else {
+        if (!hasOwn(handlerPlus, trapName)) {
+          defineProperty(handlerPlus, trapName, {
+            value: trapPlus,
+            writable: false,
+            enumerable: true,
+            configurable: true,
+          });
+        }
+        const { originalHandler } = handlerPlus;
+        const trap = originalHandler[trapName];
+        if (trap !== undefined) {
+          // Note that whether `trap === undefined` can change dynamically,
+          // so we do not install an own `handlerPlus[trapName] === undefined`
+          // for that case. We still install or preserve an own
+          // `handlerPlus[trapName] === trapPlus` until the target is
+          // seen to be non-trapping.
+          return apply(trap, originalHandler, [target, ...rest]);
+        }
+      }
+      return ReflectPlus[trapName](target, ...rest);
+    });
+    return trapPlus;
+  },
+});
+
+/**
+ * A handlerPlus starts as a fresh empty object that inherits from a proxy
+ * whose handler is the shared generic metaHandler.
+ * Thus, the metaHandler's `get` method is called only when the
+ * `handlerPlus` does not have a property overriding that `trapName`.
+ * In that case, the metaHandler's `get` is called with its `receiver`
+ * being the `handlerPlus`.
+ *
+ * @param {ProxyHandler<any>} originalHandler
+ * @returns {ProxyHandler<any> & {
+ *   isNoTrapping: (target: any) => boolean,
+ *   suppressTrapping: (target: any) => boolean,
+ *   originalHandler: ProxyHandler<any>
+ * }}
+ */
+const makeHandlerPlus = originalHandler => ({
+  // @ts-expect-error TS does not know what this __proto__ is doing
+  __proto__: new OriginalProxy({}, metaHandler),
+  // relies on there never being a trap named `originalHandler`.
+  originalHandler,
+});
+
+/**
+ * In the shim, `ProxyPlus` replaces the global `Proxy`.
+ *
+ * @type {ProxyConstructor}
+ */
+// @ts-expect-error We reject non-new calls in the body
+const ProxyPlus = function Proxy(target, handler) {
+  // @ts-expect-error Yes, we mean to compare these.
+  if (new.target !== ProxyPlus) {
+    if (new.target === undefined) {
+      throw TypeError('Proxy constructor requires "new"');
+    }
+    throw TypeError('Safe Proxy shim does not support subclassing');
+  }
+  const handlerPlus = makeHandlerPlus(handler);
+  const proxy = new OriginalProxy(target, handlerPlus);
+  proxyHandlerMap.set(proxy, [target, handler]);
+  return proxy;
+};
+// The `OriginalProxy` is both constructible (i.e., responsive to `new`) and
+// lacks a `prototype` property. The closest we can come to this is to set
+// `ProxyPlus.prototype` to `undefined`
+ProxyPlus.prototype = undefined;
```

Comment:
Not entirely true. Since we never care about the receiver in the case of a plain call, you could use `.bind(null)` which creates a function that lacks a `.prototype` but is constructible if the original is. You'll of course need to redefine the `.name`.

Reply:
Awesome, thanks! I completely missed that.

Done.

Reply:
I wouldn't expect anyone to know this, it's a very obscure quirk of `.bind()`